lib/utils/deeply-non-duplicated-addon.js

```diff
@@ -0,0 +1,44 @@
+'use strict';
+
+/**
+ * Deduplicate one addon's children addons recursively from hostAddons.
+ *
+ * @private
+ * @param {Object} hostAddons
+ * @param {EmberAddon} dedupedAddon
+ * @param {String} treeName
+ */
+module.exports = function deeplyNonDuplicatedAddon(hostAddons, dedupedAddon, treeName) {
+  if (dedupedAddon.addons.length === 0) {
+    return;
+  }
+
+  dedupedAddon._orginalAddons = dedupedAddon.addons;
+  dedupedAddon.addons = dedupedAddon.addons.filter(addon => {
+    if (addon.lazyLoading && addon.lazyLoading.enabled) {
```

Comment:
here is another place that we can use `ensureLazyLoadingHash` util as well

Reply:
I think the idea is if it encounters a child lazy engine inside a "deeplyNonDuplicatedAddon", it could just skip the resursive "deeplyNonDuplicatedAddon" for this child lazy engine for now and keep it, and when it goes to the child lazy engine's own "treeFor" method, it will pass through it's own "deeplyNonDuplicatedAddon".

For example, `father-lazy-engine-A` has `child-lazy-engine-B` and `normal-addon-C`, the "deeplyNonDuplicatedAddon" of `father-lazy-engine-A` will also recursively "deeplyNonDuplicatedAddon" `normal-addon-C` but `child-lazy-engine-B`, and then "child-lazy-engine-B" will have it's own "deeplyNonDuplicatedAddon".

I just unintentionally resovle this issue, please feel free to reopen it.

Reply:
It needs the comment indeed, it takes me sometime to recall the reason why.